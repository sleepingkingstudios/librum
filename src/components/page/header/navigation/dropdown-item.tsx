import * as React from 'react';
import {
  Link,
  useLocation,
} from 'react-router-dom';

import { NavigationItemProps } from './item';
import { useThemeStyles } from '@themes';

export const DropdownItem = ({ label, url }: NavigationItemProps): JSX.Element => {
  const className = useThemeStyles('navigation-dropdown-item');
  const linkClassName = useThemeStyles('navigation-dropdown-link');
  const disabledClassName = useThemeStyles('navigation-dropdown-link-disabled');
  const { pathname } = useLocation();

  if (pathname === url) {
    return (
      <li className={className}>
        <p className={disabledClassName}>{ label }</p>
      </li>
    );
  } else {
    return (
      <li className={className}>
        <Link className={linkClassName} to={url}>
          { label }
        </Link>
      </li>
    );
  }
};
