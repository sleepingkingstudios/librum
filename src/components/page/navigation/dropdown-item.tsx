import * as React from 'react';
import {
  Link,
  useLocation,
} from 'react-router-dom';

import type { NavigationItemProps } from './item';

export const DropdownItem = ({ label, url }: NavigationItemProps): JSX.Element => {
  const { pathname } = useLocation();

  if (pathname === url) {
    return (
      <li className="dropdown-item">
        <p className="dropdown-item-link-disabled text-muted">{ label }</p>
      </li>
    );
  } else {
    return (
      <li className="dropdown-item">
        <Link className="dropdown-item-link" to={url}>
          { label }
        </Link>
      </li>
    );
  }
};
