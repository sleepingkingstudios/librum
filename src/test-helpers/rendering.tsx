import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
