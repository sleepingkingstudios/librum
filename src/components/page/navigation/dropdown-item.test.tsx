import * as React from 'react';

import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import { DropdownItem } from './dropdown-item';
import { render } from '@test-helpers/rendering';

describe('<DropdownItem>', () => {
  const defaultProps = {
    label: 'Go Back',
    url: '/back',
  };

  describe('when the pathname does not match the url', () => {
    it('should render the link', () => {
      const { getByRole } =
        render(<DropdownItem {...defaultProps} />, { router: true });

      const link = getByRole('link', { name: defaultProps.label });

      expect(link).toBeVisible();
    });

    it('should match the snapshot', () => {
      const { asFragment } =
        render(<DropdownItem {...defaultProps} />, { router: true });

      expect(asFragment()).toMatchSnapshot();
    });

    describe('when the user clicks the link', () => {
      it('should not render a link', async () => {
        const { getByRole, getByText } =
          render(<DropdownItem {...defaultProps} />, { router: true });

        const link = getByRole('link', { name: defaultProps.label });

        await userEvent.click(link);

        expect(link).not.toBeVisible();

        expect(getByText(defaultProps.label)).toBeVisible();
      });
    });
  });

  describe('when the pathname matches the url', () => {
    it('should not render a link', () => {
      const {
        getByText,
        queryByRole,
      } = render(
        <DropdownItem {...defaultProps} />,
        {
          initialEntries: [defaultProps.url],
          router: true,
        },
      );

      expect(queryByRole('link')).toBeNull();

      expect(getByText(defaultProps.label)).toBeVisible();
    });

    it('should match the snapshot', () => {
      const { asFragment } =
        render(
          <DropdownItem {...defaultProps} />,
          {
            initialEntries: [defaultProps.url],
            router: true,
          },
        );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
