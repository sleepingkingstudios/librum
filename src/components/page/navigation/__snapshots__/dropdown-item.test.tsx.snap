// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownItem> when the pathname does not match the url should match the snapshot 1`] = `
<DocumentFragment>
  <li
    class="dropdown-item"
  >
    <a
      class="dropdown-item-link"
      href="/back"
    >
      Go Back
    </a>
  </li>
</DocumentFragment>
`;

exports[`<DropdownItem> when the pathname matches the url should match the snapshot 1`] = `
<DocumentFragment>
  <li
    class="dropdown-item"
  >
    <p
      class="dropdown-item-link-disabled text-muted"
    >
      Go Back
    </p>
  </li>
</DocumentFragment>
`;
